File: tests/test_dift.cpp
// tests/test_dift.cpp — unit tests for the DiFT torsional parametrization engine
//
// Copyright 2026 Le Bonhomme Pharma
// SPDX-License-Identifier: Apache-2.0
//
// Covers: forward transform / round-trip exactness, Shannon-collapse
// truncation, QM–MM iterative refinement, per-bond torsional thermodynamics,
// Boltzmann inversion, circular mean, and the GA scoring adapter.

#include <gtest/gtest.h>

#include "../LIB/DiFT/DiFT.h"
#include "../LIB/DiFT/DiFTGAAdapter.h"

#include <cmath>
#include <numeric>
#include <vector>

using namespace dift;

namespace {

constexpr double kTwoPi = 6.283185307179586476925286766559;

// Build an M-point sample of  V(φ) = mean + Σ Aₙ cos(nφ − ωₙ).
std::vector<double> make_profile(int M, double mean,
                                 const std::vector<int>& mult,
                                 const std::vector<double>& amp,
                                 const std::vector<double>& phase) {
    std::vector<double> p(static_cast<std::size_t>(M));
    for (int k = 0; k < M; ++k) {
        const double phi = kTwoPi * k / M;
        double v = mean;
        for (std::size_t t = 0; t < mult.size(); ++t)
            v += amp[t] * std::cos(mult[t] * phi - phase[t]);
        p[static_cast<std::size_t>(k)] = v;
    }
    return p;
}

// Look up the amplitude of a given multiplicity in a spectrum.
double amp_of(const std::vector<FourierTerm>& spec, int n) {
    for (const auto& t : spec) if (t.multiplicity == n) return t.amplitude;
    return 0.0;
}

} // namespace

// ─── forward transform: power-of-two grid (radix-2 FFT path) ─────────────────
TEST(DiFTTransform, RecoversAmplitudesPow2) {
    DiFTEngine engine(300.0);
    const int M = 64;
    auto profile = make_profile(M, 1.5, {1, 3, 5},
                                {2.0, 1.0, 0.5}, {0.0, 0.7, -1.2});
    double mean = 0.0;
    auto spec = engine.transform(profile, mean);

    EXPECT_NEAR(mean, 1.5, 1e-9);
    EXPECT_NEAR(amp_of(spec, 1), 2.0, 1e-9);
    EXPECT_NEAR(amp_of(spec, 3), 1.0, 1e-9);
    EXPECT_NEAR(amp_of(spec, 5), 0.5, 1e-9);
    EXPECT_NEAR(amp_of(spec, 2), 0.0, 1e-9);   // absent frequency
    EXPECT_NEAR(amp_of(spec, 4), 0.0, 1e-9);
}

// ─── forward transform: non-power-of-two grid (direct DFT path) ──────────────
TEST(DiFTTransform, RecoversAmplitudesNonPow2) {
    DiFTEngine engine(300.0);
    const int M = 36;                          // typical 10°-resolution scan
    auto profile = make_profile(M, -0.4, {2, 4},
                                {1.3, 0.8}, {0.3, 2.1});
    double mean = 0.0;
    auto spec = engine.transform(profile, mean);

    EXPECT_NEAR(mean, -0.4, 1e-9);
    EXPECT_NEAR(amp_of(spec, 2), 1.3, 1e-8);
    EXPECT_NEAR(amp_of(spec, 4), 0.8, 1e-8);
}

// ─── round-trip: full spectrum reconstructs the input exactly ────────────────
TEST(DiFTTransform, RoundTripReconstruction) {
    DiFTEngine engine(300.0);
    const int M = 72;
    auto profile = make_profile(M, 0.9, {1, 2, 6},
                                {1.1, 0.6, 0.25}, {-0.5, 1.0, 2.8});
    double mean = 0.0;
    auto spec = engine.transform(profile, mean);

    TorsionalPotential pot;
    pot.terms = spec;
    pot.mean  = mean;
    auto model = pot.sample(M);

    for (int k = 0; k < M; ++k)
        EXPECT_NEAR(model[static_cast<std::size_t>(k)],
                    profile[static_cast<std::size_t>(k)], 1e-7);
    EXPECT_NEAR(DiFTEngine::r_squared(profile, model), 1.0, 1e-9);
}

TEST(DiFTTransform, RejectsShortProfile) {
    DiFTEngine engine;
    std::vector<double> tiny{1.0};
    double mean = 0.0;
    EXPECT_THROW(engine.transform(tiny, mean), std::invalid_argument);
}

// ─── spectral Shannon entropy ────────────────────────────────────────────────
TEST(DiFTSpectralEntropy, SingleModeCollapsesToOne) {
    DiFTEngine engine;
    auto profile = make_profile(64, 0.0, {2}, {3.0}, {0.4});
    double mean = 0.0;
    auto spec = engine.transform(profile, mean);

    const double H = spectral_entropy(spec);
    EXPECT_NEAR(H, 0.0, 1e-9);                 // one mode → zero entropy
    EXPECT_NEAR(std::exp(H), 1.0, 1e-9);       // N_eff ≈ 1
}

TEST(DiFTSpectralEntropy, EqualPowerSpreadsToLogN) {
    // Three equal-amplitude modes → H = ln 3, N_eff = 3.
    DiFTEngine engine;
    auto profile = make_profile(64, 0.0, {1, 2, 3},
                                {1.0, 1.0, 1.0}, {0.0, 0.0, 0.0});
    double mean = 0.0;
    auto spec = engine.transform(profile, mean);
    EXPECT_NEAR(spectral_entropy(spec), std::log(3.0), 1e-9);
}

// ─── Shannon-collapse truncation in parametrize() ────────────────────────────
TEST(DiFTParametrize, KeepsDominantModesOnly) {
    DiFTEngine engine;
    // Two strong modes + a tiny ripple the collapse should discard.
    auto profile = make_profile(128, 0.2, {1, 4, 11},
                                {2.5, 1.8, 0.02}, {0.1, -0.6, 1.4});
    auto pot = engine.parametrize(profile);

    EXPECT_GE(pot.n_terms(), 2u);
    EXPECT_LE(pot.n_terms(), 3u);
    EXPECT_GT(pot.effective_modes, 1.0);
    EXPECT_LT(pot.effective_modes, 3.0);
    EXPECT_GT(pot.r_squared, 0.99);            // dominant modes capture the shape
}

TEST(DiFTParametrize, MaxMultiplicityGuard) {
    DiFTEngine engine;
    auto profile = make_profile(64, 0.0, {2, 9}, {1.0, 1.0}, {0.0, 0.0});
    auto pot = engine.parametrize(profile, /*max_multiplicity=*/6);
    for (const auto& t : pot.terms)
        EXPECT_LE(t.multiplicity, 6);
}

// ─── iterative QM–MM refinement (paper eq. 18) ───────────────────────────────
TEST(DiFTRefine, ConvergesToQMTarget) {
    DiFTEngine engine(300.0);
    const int M = 96;
    auto qm = make_profile(M, 0.0, {1, 2, 3},
                           {3.0, 1.5, 0.8}, {0.2, -0.9, 1.7});
    auto mm = make_profile(M, 0.0, {1, 2},
                           {2.0, 1.0}, {0.2, -0.9});   // missing the n=3 term

    auto corrected = engine.refine(qm, mm, /*lambda=*/0.5, /*r2_target=*/0.98);

    EXPECT_GE(corrected.r_squared, 0.98);
    EXPECT_GT(corrected.refinement_iters, 0);

    // mm + correction should now track qm.
    auto corr = corrected.sample(M);
    std::vector<double> total(static_cast<std::size_t>(M));
    for (int k = 0; k < M; ++k)
        total[static_cast<std::size_t>(k)] =
            mm[static_cast<std::size_t>(k)] + corr[static_cast<std::size_t>(k)];
    EXPECT_GE(DiFTEngine::r_squared(qm, total), 0.98);
}

TEST(DiFTRefine, RejectsMismatchedGrids) {
    DiFTEngine engine;
    std::vector<double> qm(64, 0.0), mm(32, 0.0);
    EXPECT_THROW(engine.refine(qm, mm), std::invalid_argument);
}

// ─── per-bond torsional thermodynamics ───────────────────────────────────────
TEST(DiFTThermodynamics, FreeRotorHasZeroExcessEntropy) {
    DiFTEngine engine(300.0);
    std::vector<double> flat(64, 0.0);          // V ≡ 0 → free rotor
    auto pot = engine.parametrize(flat);
    auto th  = engine.thermodynamics(pot);

    EXPECT_NEAR(th.partition_function, 1.0, 1e-6);
    EXPECT_NEAR(th.entropy,  0.0, 1e-6);
    EXPECT_NEAR(th.minus_TS, 0.0, 1e-6);
}

TEST(DiFTThermodynamics, ConfinedBondPaysEntropyPenalty) {
    DiFTEngine engine(300.0);
    // A deep single-well torsional barrier (5 kcal/mol amplitude).
    auto profile = make_profile(128, 0.0, {1}, {5.0}, {0.0});
    auto pot = engine.parametrize(profile);
    auto th  = engine.thermodynamics(pot);

    EXPECT_LT(th.partition_function, 1.0);      // confinement shrinks z
    EXPECT_LT(th.entropy,  0.0);                // entropy LOSS vs free rotor
    EXPECT_GT(th.minus_TS, 0.0);                // → positive ΔG penalty
    EXPECT_GT(th.mean_energy, 0.0);
}

TEST(DiFTThermodynamics, DeeperWellPaysMorePenalty) {
    DiFTEngine engine(300.0);
    auto shallow = engine.parametrize(
        make_profile(128, 0.0, {1}, {1.0}, {0.0}));
    auto deep = engine.parametrize(
        make_profile(128, 0.0, {1}, {6.0}, {0.0}));
    EXPECT_GT(engine.thermodynamics(deep).minus_TS,
              engine.thermodynamics(shallow).minus_TS);
}

// ─── Boltzmann inversion of a CG dihedral histogram (paper eq. 20) ───────────
TEST(DiFTBoltzmannInvert, UniformHistogramIsFlat) {
    DiFTEngine engine(300.0);
    std::vector<double> uniform(36, 100.0);
    auto energy = engine.boltzmann_invert(uniform);
    for (double e : energy) EXPECT_NEAR(e, 0.0, 1e-9);
}

TEST(DiFTBoltzmannInvert, PeakedHistogramBecomesWell) {
    DiFTEngine engine(300.0);
    std::vector<double> hist(36, 1.0);
    hist[18] = 1000.0;                          // sharp population peak
    auto energy = engine.boltzmann_invert(hist);

    EXPECT_NEAR(energy[18], 0.0, 1e-9);         // peak → energy minimum
    for (std::size_t k = 0; k < energy.size(); ++k) {
        EXPECT_GE(energy[k], -1e-12);           // all energies ≥ 0 after shift
        if (k != 18) { EXPECT_GT(energy[k], 0.0); }
    }
}

TEST(DiFTBoltzmannInvert, RejectsEmptyHistogram) {
    DiFTEngine engine;
    std::vector<double> empty(20, 0.0);
    EXPECT_THROW(engine.boltzmann_invert(empty), std::invalid_argument);
}

// ─── circular mean of phase offsets ──────────────────────────────────────────
TEST(DiFTCircularMean, HandlesWraparound) {
    // +170° and −170° straddle ±π; the circular mean is 180°, not 0°.
    std::vector<double> angles{ 170.0 * M_PI / 180.0,
                               -170.0 * M_PI / 180.0 };
    const double m = DiFTEngine::circular_mean(angles);
    EXPECT_NEAR(std::abs(m), M_PI, 1e-6);
}

TEST(DiFTCircularMean, SimpleAverage) {
    std::vector<double> angles{0.1, 0.2, 0.3};
    EXPECT_NEAR(DiFTEngine::circular_mean(angles), 0.2, 1e-6);
}

// ─── GA scoring adapter ──────────────────────────────────────────────────────
TEST(DiFTGAAdapter, ScoresTorsionalEnergyAndEntropy) {
    // Two rotatable bonds, each with a single-well DiFT potential. The π phase
    // makes V(φ) = −A·cos(nφ), so the well minimum sits at φ = 0.
    auto rbt0 = make_bond_torsion(
        make_profile(128, 0.0, {1}, {3.0}, {M_PI}), /*gene_index=*/0);
    auto rbt1 = make_bond_torsion(
        make_profile(128, 0.0, {2}, {2.0}, {M_PI}), /*gene_index=*/1);
    std::vector<RotatableBondTorsion> bonds{rbt0, rbt1};

    // Place both bonds at their well minima (φ = 0).
    std::vector<double> angles{0.0, 0.0};
    auto score = score_torsional(bonds, angles, 300.0);

    EXPECT_EQ(score.n_bonds, 2);
    EXPECT_NEAR(score.energy, 0.0, 1e-3);       // at the minima → ~0 energy
    EXPECT_GT(score.minus_TS, 0.0);             // confinement penalty > 0
    EXPECT_GT(score.total(), 0.0);
}

TEST(DiFTGAAdapter, EnergyRisesAwayFromMinimum) {
    // V(φ) = −4·cos(φ): well minimum at φ = 0, barrier crest at φ = π.
    auto rbt = make_bond_torsion(
        make_profile(128, 0.0, {1}, {4.0}, {M_PI}), 0);
    std::vector<RotatableBondTorsion> bonds{rbt};

    std::vector<double> at_min{0.0};
    std::vector<double> at_top{M_PI};
    const double e_min = score_torsional(bonds, at_min).energy;
    const double e_top = score_torsional(bonds, at_top).energy;
    EXPECT_GT(e_top, e_min + 1.0);              // barrier crest costs energy
}

TEST(DiFTGAAdapter, MismatchedInputReturnsZero) {
    auto rbt = make_bond_torsion(std::vector<double>(64, 0.0), 0);
    std::vector<RotatableBondTorsion> bonds{rbt};
    std::vector<double> angles{0.0, 0.0};       // wrong length
    auto score = score_torsional(bonds, angles);
    EXPECT_EQ(score.n_bonds, 0);
    EXPECT_EQ(score.total(), 0.0);
}

File: python/tests/test_dift.py
"""Tests for the flexaidds.dift module (pure-Python DiFT engine).

These tests verify the DiFT torsional parametrization engine — forward
transform, Shannon-collapse truncation, QM–MM refinement, per-bond
thermodynamics, Boltzmann inversion, and the GA scoring adapter — without
needing the compiled C++ extension.

SPDX-License-Identifier: Apache-2.0
"""

import math

import numpy as np
import pytest

from flexaidds.dift import (
    DiFTEngine,
    FourierTerm,
    TorsionalPotential,
    RotatableBondTorsion,
    score_torsional,
    make_bond_torsion,
    spectral_entropy,
    kB_kcal,
)

_TWO_PI = 2.0 * math.pi


# ── helpers ──────────────────────────────────────────────────────────────────

def make_profile(m, mean, mult, amp, phase):
    """M-point sample of V(φ) = mean + Σ Aₙ cos(nφ − ωₙ)."""
    k = np.arange(m, dtype=float)
    phi = k * (_TWO_PI / m)
    v = np.full(m, float(mean))
    for n, a, w in zip(mult, amp, phase):
        v += a * np.cos(n * phi - w)
    return v


def amp_of(spectrum, n):
    for t in spectrum:
        if t.multiplicity == n:
            return t.amplitude
    return 0.0


# ── forward transform ────────────────────────────────────────────────────────

def test_transform_recovers_amplitudes_pow2():
    engine = DiFTEngine(300.0)
    profile = make_profile(64, 1.5, [1, 3, 5], [2.0, 1.0, 0.5], [0.0, 0.7, -1.2])
    spectrum, mean = engine.transform(profile)
    assert mean == pytest.approx(1.5, abs=1e-9)
    assert amp_of(spectrum, 1) == pytest.approx(2.0, abs=1e-9)
    assert amp_of(spectrum, 3) == pytest.approx(1.0, abs=1e-9)
    assert amp_of(spectrum, 5) == pytest.approx(0.5, abs=1e-9)
    assert amp_of(spectrum, 2) == pytest.approx(0.0, abs=1e-9)


def test_transform_recovers_amplitudes_non_pow2():
    engine = DiFTEngine(300.0)
    profile = make_profile(36, -0.4, [2, 4], [1.3, 0.8], [0.3, 2.1])
    spectrum, mean = engine.transform(profile)
    assert mean == pytest.approx(-0.4, abs=1e-9)
    assert amp_of(spectrum, 2) == pytest.approx(1.3, abs=1e-8)
    assert amp_of(spectrum, 4) == pytest.approx(0.8, abs=1e-8)


def test_transform_round_trip():
    engine = DiFTEngine(300.0)
    m = 72
    profile = make_profile(m, 0.9, [1, 2, 6], [1.1, 0.6, 0.25], [-0.5, 1.0, 2.8])
    spectrum, mean = engine.transform(profile)
    pot = TorsionalPotential(terms=spectrum, mean=mean)
    model = pot.sample(m)
    np.testing.assert_allclose(model, profile, atol=1e-7)
    assert DiFTEngine.r_squared(profile, model) == pytest.approx(1.0, abs=1e-9)


def test_transform_rejects_short_profile():
    engine = DiFTEngine()
    with pytest.raises(ValueError):
        engine.transform([1.0])


# ── spectral Shannon entropy ─────────────────────────────────────────────────

def test_spectral_entropy_single_mode_collapses():
    engine = DiFTEngine()
    spectrum, _ = engine.transform(make_profile(64, 0.0, [2], [3.0], [0.4]))
    h = spectral_entropy(spectrum)
    assert h == pytest.approx(0.0, abs=1e-9)
    assert math.exp(h) == pytest.approx(1.0, abs=1e-9)


def test_spectral_entropy_equal_power_spreads():
    engine = DiFTEngine()
    spectrum, _ = engine.transform(
        make_profile(64, 0.0, [1, 2, 3], [1.0, 1.0, 1.0], [0.0, 0.0, 0.0]))
    assert spectral_entropy(spectrum) == pytest.approx(math.log(3.0), abs=1e-9)


# ── parametrize: Shannon-collapse truncation ─────────────────────────────────

def test_parametrize_keeps_dominant_modes():
    engine = DiFTEngine()
    profile = make_profile(128, 0.2, [1, 4, 11],
                           [2.5, 1.8, 0.02], [0.1, -0.6, 1.4])
    pot = engine.parametrize(profile)
    assert 2 <= pot.n_terms <= 3
    assert 1.0 < pot.effective_modes < 3.0
    assert pot.r_squared > 0.99


def test_parametrize_max_multiplicity_guard():
    engine = DiFTEngine()
    profile = make_profile(64, 0.0, [2, 9], [1.0, 1.0], [0.0, 0.0])
    pot = engine.parametrize(profile, max_multiplicity=6)
    assert all(t.multiplicity <= 6 for t in pot.terms)


# ── iterative QM–MM refinement ───────────────────────────────────────────────

def test_refine_converges_to_qm():
    engine = DiFTEngine(300.0)
    m = 96
    qm = make_profile(m, 0.0, [1, 2, 3], [3.0, 1.5, 0.8], [0.2, -0.9, 1.7])
    mm = make_profile(m, 0.0, [1, 2], [2.0, 1.0], [0.2, -0.9])
    corrected = engine.refine(qm, mm, lambda_=0.5, r2_target=0.98)
    assert corrected.r_squared >= 0.98
    assert corrected.refinement_iters > 0
    total = mm + corrected.sample(m)
    assert DiFTEngine.r_squared(qm, total) >= 0.98


def test_refine_rejects_mismatched_grids():
    engine = DiFTEngine()
    with pytest.raises(ValueError):
        engine.refine([0.0] * 64, [0.0] * 32)


# ── per-bond torsional thermodynamics ────────────────────────────────────────

def test_thermo_free_rotor_zero_entropy():
    engine = DiFTEngine(300.0)
    pot = engine.parametrize([0.0] * 64)
    th = engine.thermodynamics(pot)
    assert th.partition_function == pytest.approx(1.0, abs=1e-6)
    assert th.entropy == pytest.approx(0.0, abs=1e-6)
    assert th.minus_TS == pytest.approx(0.0, abs=1e-6)


def test_thermo_confined_bond_pays_penalty():
    engine = DiFTEngine(300.0)
    pot = engine.parametrize(make_profile(128, 0.0, [1], [5.0], [0.0]))
    th = engine.thermodynamics(pot)
    assert th.partition_function < 1.0
    assert th.entropy < 0.0          # entropy loss vs free rotor
    assert th.minus_TS > 0.0         # → positive ΔG penalty
    assert th.mean_energy > 0.0


def test_thermo_deeper_well_pays_more():
    engine = DiFTEngine(300.0)
    shallow = engine.parametrize(make_profile(128, 0.0, [1], [1.0], [0.0]))
    deep = engine.parametrize(make_profile(128, 0.0, [1], [6.0], [0.0]))
    assert (engine.thermodynamics(deep).minus_TS
            > engine.thermodynamics(shallow).minus_TS)


# ── Boltzmann inversion ──────────────────────────────────────────────────────

def test_boltzmann_invert_uniform_is_flat():
    engine = DiFTEngine(300.0)
    energy = engine.boltzmann_invert([100.0] * 36)
    np.testing.assert_allclose(energy, 0.0, atol=1e-9)


def test_boltzmann_invert_peaked_becomes_well():
    engine = DiFTEngine(300.0)
    hist = [1.0] * 36
    hist[18] = 1000.0
    energy = engine.boltzmann_invert(hist)
    assert energy[18] == pytest.approx(0.0, abs=1e-9)
    assert np.all(energy >= -1e-12)
    assert np.all(np.delete(energy, 18) > 0.0)


def test_boltzmann_invert_rejects_empty():
    engine = DiFTEngine()
    with pytest.raises(ValueError):
        engine.boltzmann_invert([0.0] * 20)


# ── circular mean ────────────────────────────────────────────────────────────

def test_circular_mean_wraparound():
    angles = [math.radians(170.0), math.radians(-170.0)]
    assert abs(DiFTEngine.circular_mean(angles)) == pytest.approx(math.pi, abs=1e-6)


def test_circular_mean_simple():
    assert DiFTEngine.circular_mean([0.1, 0.2, 0.3]) == pytest.approx(0.2, abs=1e-6)


# ── GA scoring adapter ───────────────────────────────────────────────────────

def test_adapter_scores_energy_and_entropy():
    rbt0 = make_bond_torsion(make_profile(128, 0.0, [1], [3.0], [math.pi]), 0)
    rbt1 = make_bond_torsion(make_profile(128, 0.0, [2], [2.0], [math.pi]), 1)
    score = score_torsional([rbt0, rbt1], [0.0, 0.0], 300.0)
    assert score.n_bonds == 2
    assert score.energy == pytest.approx(0.0, abs=1e-3)
    assert score.minus_TS > 0.0
    assert score.total() > 0.0


def test_adapter_energy_rises_away_from_minimum():
    rbt = make_bond_torsion(make_profile(128, 0.0, [1], [4.0], [math.pi]), 0)
    e_min = score_torsional([rbt], [0.0]).energy
    e_top = score_torsional([rbt], [math.pi]).energy
    assert e_top > e_min + 1.0


def test_adapter_mismatched_input_returns_zero():
    rbt = make_bond_torsion([0.0] * 64, 0)
    score = score_torsional([rbt], [0.0, 0.0])
    assert score.n_bonds == 0
    assert score.total() == 0.0


# ── cross-check: C++ ⇔ Python parity (skipped if _core lacks DiFT) ───────────

def test_cpp_python_parity_if_available():
    """If the compiled core exposes DiFTEngine, results must match Python."""
    try:
        from flexaidds import _core
        cpp_engine = _core.DiFTEngine(300.0)
    except (ImportError, AttributeError):
        pytest.skip("_core.DiFTEngine not built")

    profile = make_profile(64, 0.3, [1, 3], [2.0, 1.0], [0.4, -0.8])
    py_pot = DiFTEngine(300.0).parametrize(list(profile))
    cpp_pot = cpp_engine.parametrize(list(profile))
    assert cpp_pot.r_squared == pytest.approx(py_pot.r_squared, abs=1e-6)
    assert cpp_pot.effective_modes == pytest.approx(py_pot.effective_modes, abs=1e-6)
